### ISSUE TITLE
Depth of Field does not work on apple Devices

### ISSUE THREAD

**Issue by DonPedroo** [vanruesc/postprocessing#220]
It looks like in Safari on Apple devices (Iphone 8, Iphone X, ipad Pro) Depth of field does not work correctly, it is displaying black gradient instead of the blur effect.  Referring to this example: https://vanruesc.github.io/postprocessing/public/demo/#depth-of-field

Thanks.

**Comment by vanruesc** [comment]
Hey, thanks for reporting this. Is there any relevant output in the console? I don't own any Apple devices and can't test this.

**Comment by DonPedroo** [comment]
unfortunately you can't see anymore console log, instead, you have to connect your iphone/ipad to macbookpro with a cable  !  and I do not have any cables, and not able to change apple id since both devices must be signed under one apple id and synced in cloud. I hope maybe someone else will see this post and able to setup this crazy apple connection just to see console log.

**Comment by drcmda** [comment]
we published http://zustand.surge.sh and unfortunately it blacks out on ios.

here are the logs:

![unknown](https://user-images.githubusercontent.com/2223602/90451230-a6499d00-e0eb-11ea-835b-33c3ee96bdea.png)
![unknown2](https://user-images.githubusercontent.com/2223602/90451235-a8abf700-e0eb-11ea-9082-8d2f0a842231.png)

"too many uniforms" :-S

**Comment by vanruesc** [comment]
Thanks for those logs!

It's weird that `postprocessing` isn't logging anything, though. It [should warn you](https://github.com/vanruesc/postprocessing/blob/0dc459d17a9bbbff019af4d522aea9b9b956792d/src/passes/EffectPass.js#L368-L389) if the combined effects use too many uniforms. Any chance to get the complete logs and or the maximum supported number of uniforms on that device?

Just to confirm: the app does run on my android phone.

**Comment by giulioz** [comment]
![image](https://user-images.githubusercontent.com/1353142/90453659-7f8e6500-e0f1-11ea-9940-e61e08ea5fda.png)


```
Supported Extensions:
--
  | EXT_blend_minmax
  | EXT_sRGB
  | OES_texture_float
  | OES_texture_float_linear
  | OES_texture_half_float
  | OES_texture_half_float_linear
  | OES_standard_derivatives
  | EXT_shader_texture_lod
  | EXT_texture_filter_anisotropic
  | OES_vertex_array_object
  | OES_element_index_uint
  | WEBGL_lose_context
  | WEBKIT_WEBGL_compressed_texture_pvrtc
  | WEBGL_depth_texture
  | ANGLE_instanced_arrays
  | WEBGL_debug_renderer_info


```

**Comment by vanruesc** [comment]
Thanks @giulioz, this slims it down to the bokeh blur shader which uses a [uniform array of size 128](https://github.com/vanruesc/postprocessing/blob/0dc459d17a9bbbff019af4d522aea9b9b956792d/src/materials/glsl/bokeh/shader.frag#L9). I didn't know that this is counted as 128 individual uniforms. I'll try packing the blur kernel into `vec4` instances to reduce the number of uniforms to 32 + 4.

Now this is still above the minimum limit of 16 uniforms (see [GL_MAX_FRAGMENT_UNIFORM_VECTORS](https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glGet.xml)), but using a heavy effect like bokeh on low-end devices is not the best idea to begin with. Furthermore, packing the kernel into a data texture may degrade shader performance and requires support for float type textures which low-end devices usually don't provide.

**Comment by drcmda** [comment]
it runs surprisingly well on my low end android though. btw, are we using a depthtexture like you mentioned? i think we're just discarding alpha pixels and thereby getting the effect to work. the full source is here if that helps you track it down: https://codesandbox.io/s/xgjtc?file=/src/materials/layerMaterial.js

**Comment by vanruesc** [comment]
@drcmda Yes, the DoF effect is using a depth texture attachment. The layer material you're using performs alpha testing which uses the `discard` operation. Shaders that use `discard` or write to `gl_FragDepth` disable early depth tests (check this [SO post](https://stackoverflow.com/questions/19819376/what-happens-to-the-depth-buffer-if-i-discard-a-fragment-in-a-shader-using-early) for some details). Alpha tests can kill performance on mobile devices, but it should be fine in your use case, because you don't have lots of objects that could benefit from early depth tests.

> it runs surprisingly well on my low end android though

That's good to know 👍 I based my assumption on the fact that tile-based mobile graphics processors usually struggle with convolution shaders.

**Comment by vanruesc** [comment]
Should be fixed in `postprocessing@6.17.0`.
